### ISSUE TITLE
async_pipe move assignment operator throws an exception in Windows

### ISSUE THREAD

**Issue by zhassan** [klemens-morgenstern/boost-process#189]
The following code works fine on linux but produces an exception in Windows:

```
#include <iostream>
#include <boost/asio/io_service.hpp>
#include <boost/process.hpp>

namespace bp = boost::process;

class Ap {
public:
  Ap() : m_pipe(m_service) { }

  void run() {
    m_service.reset();
    //m_pipe.close();
    m_pipe = bp::async_pipe(m_service);
    std::vector<std::string> args { "-c", "for i in 1 2 3 4 5; do echo iter $i; sleep 1; done" };
    m_process = bp::child(bp::search_path("sh"), args, bp::std_out > m_pipe);
    readOutput();
  } 
    
  int wait() {
    while (!m_process.wait_for(std::chrono::milliseconds(10))) {
      m_service.poll();
    }
    return m_process.exit_code();
  } 

private:
  boost::asio::io_service         m_service;
  bp::child                       m_process;
  bp::async_pipe                  m_pipe;
  std::array<char, 1024>          m_buf;
  
  void readOutput() {
    m_pipe.async_read_some(boost::asio::buffer(m_buf),
                           [this](const boost::system::error_code& ec, size_t size) {
                             if (!ec) {
                               std::cout << "### Output: " << m_buf.data();
                               readOutput();
                             }
                           });
  }
};

int main()
{ 
  Ap a;
  for (int i = 0; i < 2; ++i) {
    std::cout << "====================== Iter " << i << std::endl;
    a.run();
    int status = a.wait();
    std::cout << "Process exited with status " << status << std::endl;
  } 
  return 0;
}   
```
On linux, I get the following output:
```
====================== Iter 0
### Output: iter 1
### Output: iter 2
### Output: iter 3
### Output: iter 4
### Output: iter 5
Process exited with status 0
====================== Iter 1
### Output: iter 1
### Output: iter 2
### Output: iter 3
### Output: iter 4
### Output: iter 5
Process exited with status 0
```
But on Windows, I'm getting:
```
====================== Iter 0

This application has requested the Runtime to terminate it in an unusual way.
Please contact the application's support team for more information.
terminate called after throwing an instance of 'boost::exception_detail::clone_impl<boost::exception_detail::error_info_injector<boost::system::system_error> >'
  what():  assign: Already open
```
I tried closing the pipe before creating a new one (uncomment second line of Ap::run), but this results in a different exception getting thrown:
```
====================== Iter 0

This application has requested the Runtime to terminate it in an unusual way.
Please contact the application's support team for more information.
terminate called after throwing an instance of 'boost::exception_detail::clone_impl<boost::exception_detail::error_info_injector<boost::system::system_error> >'
  what():  assign: The parameter is incorrect
```
I'm using boost 1.67 and using mingw to compile for Windows.

I worked around the issue by using a unique_ptr for the async_pipe object.

**Comment by klemens-morgenstern** [comment]
Thanks for reporting. I remember this issue, it should be fixed since 1.68 or 1.69.

**Comment by zhassan** [comment]
I just tried 1.69 and I got the same error. Was the fix in the boost process library itself or in some other boost library?
